vm/src/stdlib/collections.rs

```diff
@@ -127,11 +128,22 @@ mod _collections {
         }
 
         #[pymethod]
-        fn extend(&self, iter: PyIterable, vm: &VirtualMachine) -> PyResult<()> {
+        fn extend(zelf: PyRef<Self>, iter: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
             // TODO: use length_hint here and for extendleft
-            for elem in iter.iter(vm)? {
-                self.append(elem?);
+            let max_len = zelf.maxlen.load();
+            let mut elements: Vec<PyObjectRef> = vm.extract_elements(&iter)?;
+            if let Some(max_len) = max_len {
+                let cut_len = max_len as isize - elements.len() as isize;
+                if cut_len > 0 {
+                    let mut deque = zelf.borrow_deque_mut();
+                    let drain_until = deque.len().saturating_sub(cut_len as usize);
+                    deque.drain(..drain_until);
+                } else {
+                    zelf.borrow_deque_mut().clear();
+                    elements.drain(..(-cut_len) as usize);
```

Comment:
I'd just compare these:

```suggestion
                if max_len > elements.len() {
                    let mut deque = zelf.borrow_deque_mut();
                    let drain_until = deque.len().saturating_sub(max_len - elements.len());
                    deque.drain(..drain_until);
                } else {
                    zelf.borrow_deque_mut().clear();
                    elements.drain(..elements.len() - max_len);
```

Reply:
the new code has additional subtraction operations. idk if they will be optimized out, but why if we can avoid it explicitly?

Reply:
Large values of `maxlen` (i.e `deque([1], maxlen = maxsize + 2)` cause panics. I just noticed that `maxsize` shouldn't be allowed to be `> isize::MAX`, though, and I'll fix that at another PR.

Either way, using a comparison here means we only operate on positive numbers, which means easier to follow and harder to make mistakes (personally, at least!)